File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        left, right = 0, n - 1
        leftmax, rightmax = 0, 0
        water = 0
        while left < right:
            if height[left] < height[right]:
                if height[left] >= leftmax:
                    leftmax = height[left]
                else:
                    water += leftmax - height[left]
                left += 1
            else:
                if height[right] >= rightmax:
                    rightmax = height[right]
                else:
                    water += rightmax - height[right]
                right -= 1
        return water

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length, i = 0, j = n-1, leftmost = 0, rightmost = 0, water = 0;
         while(i<j){
            if(height[i]<height[j]){
                if(height[i]>=leftmost){
                    leftmost=height[i];
                }
                else{
                    water+=leftmost-height[i];
                }
                i+=1;
            }
            else{
                if(height[j]>=rightmost){
                    rightmost=height[j];
                }
                else{
                    water+=rightmost-height[j];
                }
                j-=1;
            }
         }
         return water;
    }
}